### DISCUSSION TITLE
How to use flask-socketIo with a threaded flask application

### DISCUSSION THREAD

**Discussion by MathiasDrapier** [miguelgrinberg/Flask-SocketIO#1601] [Q&A]
Hello, I am using flask-socketio to add socketio functionality in my flask api.
Until now everything worked fine by initializing the application like this:

main.py : 

`socketio.run(app, host='::', port=1234)`

app.py:

`socketio = SocketIO(app, always_connect=True)`

But I noticed that my application could not answer several requests at the same time. 
So I initialized socketIo like this: 

`socketio = SocketIO(app, always_connect=True, async_mode="threading")`

This time my API is able to respond to several requests at the same time but when I start my application I receive the following log: 

`WebSocket transport not available. Install eventlet or gevent and gevent-websocket for improved performance.`

And my Socketio client gets disconnected 5-6 seconds after logging in.

I searched here and couldn't find a solution except removing the asyn_mode = threading and putting sleeps in the long requests but I don't want to do this. My application should not be able to manage a lot of customers so even quick requests should be able to be requested at the same time.

**Comment by bluthen** [comment]
Perhaps #1473 answers your question?

**Comment by MathiasDrapier** [comment]
Thank you for the help but I have already seen this topic and it does not really meet my expectations ... I want to run my application in a web server like eventlet for example but supporting threading (several requests at the same time ) ...

**Comment by miguelgrinberg** [comment]
@MathiasDrapier I think you need to read up on how eventlet and other async frameworks achieve concurrency.

> I want to run my application in a web server like eventlet for example but supporting threading

Eventlet does not support threads. It supports greenlets instead.

> (several requests at the same time)

It can do this easily, and actually can scale much better than threads, as long as you don't block.

> putting sleeps in the long requests but I don't want to do this

I did not make the rules. That's how eventlet works. If you don't like it, then you have to use regular threads and pay a performance penalty. I'm working on support for WebSocket in threading mode, so this is going to be available in the near future, but right now you have no other choices, so you need to decide which option you hate less. Sorry.

**Comment by MathiasDrapier** [comment]
Thank you for your response @miguelgrinberg . Is it possible with Eventlet to answer several requests at the same time? If so, how to do it?

**Comment by miguelgrinberg** [comment]
@MathiasDrapier I mentioned it above. You are likely running blocking code in your handlers. Once you stop doing that, concurrency will work.

**Comment by MathiasDrapier** [comment]
What do you mean as the blocking code ? 
Do you think I have blocking code in one of my requests from my flask api?
So I have to remove this code and remove the async_mode = "treading" option, do I have to do something else to make eventlet reach the concurrency?

**Comment by miguelgrinberg** [comment]
@MathiasDrapier you are trying to work with something you are not familiar with. I suggested above that if you intend to use eventlet you need to familiarize yourself with greenlets and how they achieve concurrency. You really need to understand what code is blocking under eventlet and how to avoid it. And if you don't want to make an investment in learning yet another thing, then go back to the threading mode and things will work better, just a bit slower and without WebSocket for now.

**Comment by MathiasDrapier** [comment]
Ok thank you, but with threading mode my socketio client is disconnected a few seconds after its connection.

**Comment by miguelgrinberg** [comment]
I need to see logs of that. See the troubleshooting section of the docs if you don't know how to get them.

**Comment by MathiasDrapier** [comment]
Here is my logs : 
```
| Server initialized for threading.

| WebSocket transport not available. Install eventlet or gevent and gevent-websocket for improved performance.
|  * Serving Flask app "app" (lazy loading)
|  * Environment: production
|    WARNING: This is a development server. Do not use it in a production deployment.
|    Use a production WSGI server instead.
|  * Debug mode: off
|  * Running on http://[::]:xxxx/ (Press CTRL+C to quit)
| hqy63yuCWaHvxE2SAAAA: Sending packet OPEN data {'sid': 'hqy63yuCWaHvxE2SAAAA', 'upgrades': [], 'pingTimeout': 20000, 'pingInterval': 25000}
| ::ffff:xx.xxx.x.x - - [19/Apr/2021 10:48:33] "GET /socket.io/?EIO=4&transport=polling HTTP/1.1" 200 -
| hqy63yuCWaHvxE2SAAAA: Received packet MESSAGE data 0
| hqy63yuCWaHvxE2SAAAA: Sending packet MESSAGE data 0{"sid":"rQAaMBnuCPUOs22uAAAB"}
| 2021-04-19 10:48:34 INFO websocket.py headers : Accept: */*
| Authorization: Bearer xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
| Connection: Keep-Alive
| Accept-Encoding: gzip
| User-Agent: okhttp/3.12.12
|
|
| ::ffff:x.x.x.x - - [19/Apr/2021 10:48:34] "GET /socket.io/?EIO=4&sid=hqy63yuCWaHvxE2SAAAA&transport=polling HTTP/1.1" 200 -
| rQAaMBnuCPUOs22uAAAB is entering room 15 [/]
| emitting event "message" to 15 [/]
| hqy63yuCWaHvxE2SAAAA: Sending packet MESSAGE data 2["message","{'type': 'xxxx'}"]
| ::ffff:x.x.x.x - - [19/Apr/2021 10:48:34] "POST /socket.io/?EIO=4&sid=hqy63yuCWaHvxE2SAAAA&transport=polling HTTP/1.1" 200 -
| ::ffff:x.x.x.x - - [19/Apr/2021 10:48:34] "GET /frontendCommunication/messages HTTP/1.1" 200 -
| ::ffff:x.x.x.x - - [19/Apr/2021 10:48:34] "GET /socket.io/?EIO=4&sid=hqy63yuCWaHvxE2SAAAA&transport=polling HTTP/1.1" 200 -
| 10722230
| hqy63yuCWaHvxE2SAAAA: Received packet CLOSE data
| emitting event "message" to 15 [/]
| hqy63yuCWaHvxE2SAAAA: Sending packet MESSAGE data 2["message","{'type': 'xxxxxx'}"]
| ::ffff:x.x.x.x - - [19/Apr/2021 10:48:44] "GET /socket.io/?EIO=4&sid=hqy63yuCWaHvxE2SAAAA&transport=polling HTTP/1.1" 200 -
| rQAaMBnuCPUOs22uAAAB is leaving room rQAaMBnuCPUOs22uAAAB [/]
| rQAaMBnuCPUOs22uAAAB is leaving room 15 [/]
| ::ffff:x.x.x.x - - [19/Apr/2021 10:48:44] "POST /socket.io/?EIO=4&sid=hqy63yuCWaHvxE2SAAAA&transport=polling HTTP/1.1" 200 -
```

**Comment by miguelgrinberg** [comment]
Your client is closing the connection:

    hqy63yuCWaHvxE2SAAAA: Received packet CLOSE data

**Comment by MathiasDrapier** [comment]
Okay sorry, I will check it. And concerning the use of eventlet, I will wait and document myself on this subject. Thank you very much for the relevant and quick response.

**Comment by MathiasDrapier** [comment]
What is the transport used if it is not the websocket? I must specify it on the client side to allow it to connect.

**Comment by miguelgrinberg** [comment]
Polling. This is the default, no need to specify it.

**Comment by MathiasDrapier** [comment]
Okay thanks

**Comment by elvis-epx** [comment]
I did something similar, I wanted to run Flask inside a larger app, so I run it in its separate thread, hoping that the gevent machinery also lives inside that thread.

Run into a problem that WebSocket messages were being delayed, and delivered in small batches instead of immediately. My conclusion was that Flask was sleeping and socketio.send() was not waking the thread up.

I conocted a crude ISP mechanism with UDP packets, just to wake up gevent loop within the working thread. The listening socket is created and monitored by start_background_task() using gevent version of socket methods.

```
...
from gevent import socket as gsocket
import socket

# initialization... 

        self.socketio = SocketIO(self.app, async_mode='gevent')

        self.wpipe = socket.socket(socket.AF_INET, socket.SOCK_DGRAM, socket.IPPROTO_UDP)
        self.wpipe.bind(('127.0.0.1', 0))
        self.wpipe.connect(('127.0.0.1', listen_port))

        def f():
            print("### bg task")
            rpipe = gsocket.socket(socket.AF_INET, socket.SOCK_DGRAM, socket.IPPROTO_UDP)
            rpipe.bind(('127.0.0.1', listen_port))
            while rpipe.recv(1514):
                print("### bg task ping")

        def g():
            self.socketio.start_background_task(f)
            self.socketio.run(self.app, host=listen_addr, port=listen_port, debug=True, use_reloader=False)
            
       self.t = Thread(daemon=True, target=g).start()

# when message is sent...
       
    def websockets_pub(self, room, msg):
        if type(msg) is not str:
            msg = json.dumps(msg)
        self.socketio.send(msg, to=room)
        self.wpipe.send(b'ping')
```
